src/io.rs

```diff
@@ -33,3 +35,135 @@ impl std::error::Error for EspIOError {}
 pub mod asynch {
     pub use embedded_io_async::*;
 }
+
+/// A driver for bufferd / blocking communication ontop of Standard IO Streams.
+///
+/// By default all communication via std::io:stdin / stdout is non-blocking.
+/// That means if a user wants to read from stdin, they have to constantly poll the driver, since
+/// the respective hardware FIFO buffers are relative small. Also users would have to handle WouldBlock
+/// errors on every call, thouse make it unergonomic to handle.
+///
+/// A BlockingIoDriver will instruct the VFS(Virtual File System) to use the drivers interrupt driven,
+/// blocking read and write functions instead.
+pub struct BlockingIoDriver<T> {
+    phantom: PhantomData<T>,
+}
+
+pub struct BlockingUartIo<T: Uart> {
```

Comment:
The `BlockingUartIo<BlockingIoDriver<T>>` indirection is unnecessary. You only need one single struct, not two. Then, depending what peripheral the user has selected with `CONFIG_` (if any), you can have different implementation code of the struct methods.

But then again, I'm not sure this pattern is worth it. Perhaps switch a a global function (or two global functions, if we expose the dev console too), and then conditionalize these depending on the peripheral selected at compile-time. And don't expose the function(s) if user has selected `CONFIG_*_NONE`.